src/dune_lang/decoder.ml

```diff
@@ -286,19 +281,33 @@ let enter t =
         result ctx (t ctx l)
       | sexp -> User_error.raise ~loc:(Ast.loc sexp) [ Pp.text "List expected" ])
 
-let if_list ~then_ ~else_ =
-  peek_exn >>= function
-  | List _ -> then_
-  | _ -> else_
-
-let if_paren_colon_form ~then_ ~else_ =
-  peek_exn >>= function
-  | List (_, Atom (loc, A s) :: _) when String.is_prefix s ~prefix:":" ->
-    let name = String.drop s 1 in
-    enter
-      ( junk >>= fun () ->
-        then_ >>| fun f -> f (loc, name) )
-  | _ -> else_
+let ( <|> ) =
+  (* Before you read this code, close your eyes and internalise the fact that
+     this code is temporary. It is a temporary state as part of a larger work to
+     turn [Decoder.t] into a pure applicative. Once this is done, this function
+     will be implemented in a better way and with a much cleaner semantic. *)
+  let approximate_how_much_input_a_failing_branch_consumed
```

Comment:
Just a note: I think parsers often disallow alternatives with overlapping prefixes to make sure there is no deep backtracking. If you do this too, then a failing parser can consume at most one symbol from the input.

Reply:
That seems like a good model indeed. Thanks for mentioning this! I'll do it once the transition is over.